Functions w/ Only Multiple Trailing Closures Trigger "No Space in Method Call" (Swift 5.3)

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

When using Swift 5.3's multiple trailing closure syntax (disabling the rule for `multiple_closures_with_trailing_closure`), a function call with *only* multiple trailing closures as arguments will trigger "No Space in Method Call Violation."

Common examples of such functions are:

1. SwiftUI Button initializer:

```swift
Button { // No Space in Method Call Violation
    print("Button tapped")
} label: {
    Text("Button")
}
```

2. Combine `sink` Operator:

```swift
Just(1)
    .sink { completion in // No Space in Method Call Violation
        print(completion)
    } receiveValue: { value in
        print(value)
    }
    .store(in: &disposeBag)
```

Function calls that begin with non-closure parameters followed by multiple trailing closures work just fine. For example:

```swift
UIView.animate(withDuration: 0.5) { // All clear, no violation
    print("animating")
} completion: { _ in
    print("done")
}
```

Removing the space properly triggers "Opening Brace Spacing Violation" (see `swiftlint autocorrect` output below).

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
Loading configuration from '.swiftlint.yml'
Linting Swift files at paths
Linting 'ContentView.swift' (1/2)
Linting 'LintTestApp.swift' (2/2)
.../LintTest/ContentView.swift:17:15: warning: No Space in Method Call Violation: Don't add a space between the method name and the parentheses. (no_space_in_method_call)
Done linting! Found 1 violations, 0 serious in 2 files.
```

```bash
$ swiftlint autocorrect
Loading configuration from '.swiftlint.yml'
Correcting Swift files at paths
Correcting 'AnimationView.swift' (1/4)
Correcting 'LintViewModel.swift' (2/4)
Correcting 'ContentView.swift' (3/4)
Correcting 'LintTestApp.swift' (4/4)
.../LintTest/LintTest/ContentView.swift:17:15 Corrected No Space in Method Call
.../LintTest/LintTest/ContentView.swift:17:15 Corrected Opening Brace Spacing
.../LintTest/LintTest/LintViewModel.swift:20:18 Corrected No Space in Method Call
.../LintTest/LintTest/LintViewModel.swift:20:18 Corrected Opening Brace Spacing
Done correcting 4 files!
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 
`0.39.2`
* Installation method used (Homebrew, CocoaPods, building from source, etc)?
Homebrew
* Paste your configuration file:

```yml
disabled_rules:
  - multiple_closures_with_trailing_closure

# Implicitly relevant:
# - no_space_in_method_call
# - opening_brace
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
No.
* Which Xcode version are you using (check `xcodebuild -version`)?
Xcode 12.0
Build version 12A6159
(Beta 1)
* Do you have a sample that shows the issue?

```swift
Button { // No Space in Method Call Violation
    print("Button tapped")
} label: {
    Text("Button")
}
```

### Additional Comments

Looking into fixing this by adjusting [NoSpaceInMethodCallRule.swift](https://github.com/realm/SwiftLint/blob/075312af6e8b2ac527a02a725a51e55f65816a0f/Source/SwiftLintFramework/Rules/Style/NoSpaceInMethodCallRule.swift#L78). Once the issue is resolved, it may be worth considering disabling the `multiple_closures_with_trailing_closure` rule by default. While the new syntax isn't universally loved, it mitigates the call site ambiguity that motivated the rule's creation.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -49,6 +49,11 @@
 * Fix false positive uppercase enum case in `raw_value_for_camel_cased_codable_enum` rule
   [Teameh](https://github.com/teameh)
 
+* Fix false positive in `no_space_in_method_call` rule with multiple trailing 
+  closures (Swift 5.3).  
+  [Marcelo Fabri](https://github.com/marcelofabri)
+  [#3259](https://github.com/realm/SwiftLint/issues/3259)
+
 * Fix false negative in `explicit_acl` rule when using `extension` with 
   Swift 5.2+.  
   [Marcelo Fabri](https://github.com/marcelofabri)

--- a/Source/SwiftLintFramework/Models/SwiftVersion.swift
+++ b/Source/SwiftLintFramework/Models/SwiftVersion.swift
@@ -31,6 +31,8 @@ public extension SwiftVersion {
     static let fiveDotOne = SwiftVersion(rawValue: "5.1.0")
     /// Swift 5.2.x - https://swift.org/download/#swift-52
     static let fiveDotTwo = SwiftVersion(rawValue: "5.2.0")
+    /// Swift 5.3.x - https://swift.org/download/#swift-53
+    static let fiveDotThree = SwiftVersion(rawValue: "5.3.0")
 
     /// The current detected Swift compiler version, based on the currently accessible SourceKit version.
     ///

--- a/Source/SwiftLintFramework/Rules/Style/NoSpaceInMethodCallRule.swift
+++ b/Source/SwiftLintFramework/Rules/Style/NoSpaceInMethodCallRule.swift
@@ -13,22 +13,15 @@ public struct NoSpaceInMethodCallRule: SubstitutionCorrectableASTRule, Configura
         description: "Don't add a space between the method name and the parentheses.",
         kind: .style,
         minSwiftVersion: .fourDotTwo,
-        nonTriggeringExamples: [
-            Example("foo()"),
-            Example("object.foo()"),
-            Example("object.foo(1)"),
-            Example("object.foo(value: 1)"),
-            Example("object.foo { print($0 }"),
-            Example("list.sorted { $0.0 < $1.0 }.map { $0.value }"),
-            Example("self.init(rgb: (Int) (colorInt))")
-        ],
+        nonTriggeringExamples: nonTriggeringExamples,
         triggeringExamples: [
             Example("foo↓ ()"),
             Example("object.foo↓ ()"),
             Example("object.foo↓ (1)"),
             Example("object.foo↓ (value: 1)"),
             Example("object.foo↓ () {}"),
-            Example("object.foo↓     ()")
+            Example("object.foo↓     ()"),
+            Example("object.foo↓     (value: 1) { x in print(x) }")
         ],
         corrections: [
             Example("foo↓ ()"): Example("foo()"),
@@ -39,6 +32,34 @@ public struct NoSpaceInMethodCallRule: SubstitutionCorrectableASTRule, Configura
             Example("object.foo↓     ()"): Example("object.foo()")
         ]
     )
+
+    private static var nonTriggeringExamples: [Example] {
+        let commonExamples = [
+            Example("foo()"),
+            Example("object.foo()"),
+            Example("object.foo(1)"),
+            Example("object.foo(value: 1)"),
+            Example("object.foo { print($0 }"),
+            Example("list.sorted { $0.0 < $1.0 }.map { $0.value }"),
+            Example("self.init(rgb: (Int) (colorInt))")
+        ]
+
+        guard SwiftVersion.current >= .fiveDotThree else {
+            return commonExamples
+        }
+
+        let swiftFiveDot3Examples = [
+            Example("""
+            Button {
+                print("Button tapped")
+            } label: {
+                Text("Button")
+            }
+            """)
+        ]
+
+        return commonExamples + swiftFiveDot3Examples
+    }
 
     // MARK: - ASTRule
 
@@ -76,9 +97,7 @@ public struct NoSpaceInMethodCallRule: SubstitutionCorrectableASTRule, Configura
         }
 
         // Don't trigger if it's a single parameter trailing closure without parens
-        if let subDict = dictionary.substructure.last,
-            subDict.expressionKind == .closure,
-            let closureBodyOffset = subDict.bodyOffset,
+        if let closureBodyOffset = dictionary.substructure.lazy.compactMap({ $0.closureBodyOffset }).first,
             closureBodyOffset == bodyOffset {
             return []
         }
@@ -89,5 +108,19 @@ public struct NoSpaceInMethodCallRule: SubstitutionCorrectableASTRule, Configura
         }
 
         return [range]
+    }
+}
+
+private extension SourceKittenDictionary {
+    var closureBodyOffset: ByteCount? {
+        if expressionKind == .closure {
+            return bodyOffset
+        }
+
+        if expressionKind == .argument, substructure.last?.expressionKind == .closure {
+            return substructure.last?.bodyOffset
+        }
+
+        return nil
     }
 }

